# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/events/PointerEventHandler.h
# Commit: 0c3a2df68be4
# Full Hash: 0c3a2df68be4e59a46e89640b95e3f8ba7082a60
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2024-03-12 09:25:52
# Regressor Bug: 1695574
# File Overlap Count: 1
# Description:
#   Bug 1695574 - Make `PresShell::EventHandler::HandleEventUsingCoordinates` keep handling touch events after preceding pointer event target is removed r=smaug,edgar,dom-core
#   
#   As far as investigating the behavior of the other browsers, touch events are
#   dispatched the same target as the preceding `pointerdown` even after the target
#   is removed from the DOM tree.  However, the following pointer events are
# ==============================================================================

diff -r bdee0bed3f5c -r 0c3a2df68be4 dom/events/PointerEventHandler.h
--- a/dom/events/PointerEventHandler.h	Tue Mar 12 00:54:41 2024 +0000
+++ b/dom/events/PointerEventHandler.h	Tue Mar 12 01:03:44 2024 +0000
@@ -168,11 +168,39 @@
   static void PostHandlePointerEventsPreventDefault(
       WidgetPointerEvent* aPointerEvent, WidgetGUIEvent* aMouseOrTouchEvent);
 
-  MOZ_CAN_RUN_SCRIPT
-  static void DispatchPointerFromMouseOrTouch(
-      PresShell* aShell, nsIFrame* aFrame, nsIContent* aContent,
-      WidgetGUIEvent* aEvent, bool aDontRetargetEvents, nsEventStatus* aStatus,
-      nsIContent** aTargetContent);
+  /**
+   * Dispatch a pointer event for aMouseOrTouchEvent to aEventTargetContent.
+   *
+   * @param aShell              The PresShell which is handling the event.
+   * @param aEventTargetFrame   The frame for aEventTargetContent.
+   * @param aEventTargetContent The event target node.
+   * @param aMouseOrTouchEvent  A mouse or touch event.
+   * @param aDontRetargetEvents If true, this won't dispatch event with
+   *                            different PresShell from aShell.  Otherwise,
+   *                            pointer events may be fired on different
+   *                            document if and only if aMouseOrTOuchEvent is a
+   *                            touch event except eTouchStart.
+   * @param aState              [out] The result of the pointer event.
+   * @param aMouseOrTouchEventTarget
+   *                            [out] The event target for the following mouse
+   *                            or touch event. If aEventTargetContent has not
+   *                            been removed from the tree, this is always set
+   *                            to it. If aEventTargetContent is removed from
+   *                            the tree and aMouseOrTouchEvent is a mouse
+   *                            event, this is set to inclusive ancestor of
+   *                            aEventTargetContent which is still connected.
+   *                            If aEventTargetContent is removed from the tree
+   *                            and aMouseOrTouchEvent is a touch event, this is
+   *                            set to aEventTargetContent because touch event
+   *                            should be dispatched even on disconnected node.
+   *                            FIXME: If the event is a touch event but the
+   *                            message is not eTouchStart, this won't be set.
+   */
+  MOZ_CAN_RUN_SCRIPT static void DispatchPointerFromMouseOrTouch(
+      PresShell* aShell, nsIFrame* aEventTargetFrame,
+      nsIContent* aEventTargetContent, WidgetGUIEvent* aMouseOrTouchEvent,
+      bool aDontRetargetEvents, nsEventStatus* aStatus,
+      nsIContent** aMouseOrTouchEventTarget = nullptr);
 
   static void InitPointerEventFromMouse(WidgetPointerEvent* aPointerEvent,
                                         WidgetMouseEvent* aMouseEvent,